twitter_scrooge/twitter_scrooge.bzl

```diff
@@ -0,0 +1,269 @@
+  )
+
+def _collect_transitive_srcs(targets):
+  r = set()
+  for target in targets:
+    if hasattr(target, "thrift"):
+      r += target.thrift.transitive_srcs
+  return r
+
+def _collect_owned_srcs(targets):
+  r = set()
+  for _target in targets:
+    if hasattr(_target, "extra_information"):
+      for target in _target.extra_information:
+        if hasattr(target, "scrooge_srcjar"):
+          r += target.scrooge_srcjar.transitive_owned_srcs
+  return r
+
+def collect_extra_srcjars(targets):
+  srcjars = set()
+  for target in targets:
+    if hasattr(target, "extra_information"):
+      for _target in target.extra_information:
+        srcjars += [_target.srcjars.srcjar]
+        srcjars += _target.srcjars.transitive_srcjars
+  return srcjars
+
+def _collect_immediate_srcs(targets):
+  r = set()
+  for target in targets:
+    if hasattr(target, "thrift"):
+      r += [target.thrift.srcs]
+  return r
+
+def _assert_set_is_subset(left, right):
+  missing = set()
+  for l in left:
+    if l not in right:
+      missing += [l]
+  if len(missing) > 0:
+    fail('scrooge_srcjar target must depend on scrooge_srcjar targets sufficient to ' +
+         'cover the transitive graph of thrift files. Uncovered sources: ' + missing)
+
+def _content_newline(data):
+  return '\n'.join([f.path for f in data])
+
+def _gen_scrooge_srcjar_impl(ctx):
+  remote_jars = set()
+  for target in ctx.attr.remote_jars:
+    remote_jars += _jar_filetype.filter(target.files)
+
+  # These are the thrift sources whose generated code we will "own" as a target
+  immediate_thrift_srcs = _collect_immediate_srcs(ctx.attr.deps)
+
+  # This is the set of sources which is covered by any scala_library
+  # or scala_scrooge_gen targets that are depended on by this. This is
+  # necessary as we only compile the sources we own, and rely on other
+  # targets compiling the rest (for the benefit of caching and correctness).
+  transitive_owned_srcs = _collect_owned_srcs(ctx.attr.deps)
+
+  # These are the thrift sources in the dependency graph. They are necessary
+  # to generate the code, but are not "owned" by this target and will not
+  # be in the resultant source jar
+
+  transitive_thrift_srcs = transitive_owned_srcs + _collect_transitive_srcs(ctx.attr.deps)
+
+  only_transitive_thrift_srcs = set()
+  for src in transitive_thrift_srcs:
+    if src not in immediate_thrift_srcs:
+      only_transitive_thrift_srcs += [src]
+
+  # We want to ensure that the thrift sources which we do not own (but need
+  # in order to generate code) have targets which will compile them.
+  _assert_set_is_subset(only_transitive_thrift_srcs, transitive_owned_srcs)
+
+  #TODO would there be a case in which we want/need to allow more jars here?
+  transitive_cjars = [ctx.files._scrooge_core, ctx.files._libthrift, ctx.files._scalalib]
+  transitive_cjars = [file for files in transitive_cjars for file in files]
+
+  cjars = ctx.files._scrooge_generator + \
```

Comment:
it was necessary when I was calling scrooge manually, but it may not be necessary now that we have a binary with scrooge's deps there...I will clean up